app.py

```diff
@@ -1,7 +1,141 @@
-from flask import Flask, render_template
-app = Flask(__name__)
+"""
+A Flask application for snapcraft.io.
+
+The web frontend for the snap store.
+"""
+
+import flask
+import requests
+import requests_cache
+import datetime
+import json
+import humanize
+from dateutil import parser
+
+
+app = flask.Flask(__name__)
+uncached_session = requests.Session()
+cached_session = requests_cache.CachedSession(expire_after=60)
+request_timeout = 2
+snap_details_url = (
+    "https://api.snapcraft.io/api/v1/snaps/details/{snap_name}"
+    "?channel=stable"
+)
 
 
 @app.errorhandler(404)
-def page_not_found(e):
-    return render_template('404.html'), 404
+def page_not_found(error):
+    """
+    For 404 pages, display the 404.html template,
+    passing through the error description.
+    """
+
+    return flask.render_template(
+        '404.html', description=error.description
+    ), 404
+
+
+@app.route('/<snap_name>/')
+def snap_details(snap_name):
+    """
+    A view to display the snap details page for specific snaps.
+
+    This queries the snapcraft API (api.snapcraft.io) and passes
+    some of the data through to the snap-details.html template,
+    with appropriate sanitation.
+    """
+
+    query_headers = {
+        'X-Ubuntu-Series': '16',
+    }
+
+    response = _get_from_cache(
```

Comment:
We discussed this briefly in our call today. I would be more concerned about retrying failed requests than caching results in the context. Sure, the load will be not neglectable on the store services but we are in a better position to cache it properly, specially when authentication comes to play. Also, considering the possible topology of these services, both deployed in PS4.5, request round-trips won't be expensive.
OTOH, retrying 5xx request has proven to be relevant to avoid network glitches to prevent external request to be fulfilled.

Reply:
I like the idea of retrying requests - that certainly makes sense. But I'd rather retry for a set amount of time (2 seconds, probably) rather than a set number of times, as I want to avoid this application hanging on a delayed API response.

I still have concerns about not providing some redundancy in this layer. For these public API requests, it seems to me to be perfectly valid to cache the response, and I think the public facing details pages should be especially resilient.

If it doesn't make sense to cache the authenticated API responses, then that's fine - I think an authenticated user would be less surprised to find that an API call was required in their requests - and it would be slightly less impactful if those pages broke.

I'd be happy to set up a different cache which honours caching headers (probably https://pypi.python.org/pypi/CacheControl) so that your application can choose how the response is cached if that would help. But for public pages at least, I'd still like to return cached data rather than a 500 page if we have an API error. We will still display a warning notification to explain that the API response failed.

Reply:
Discussion about caching moved to https://github.com/canonical-websites/snapcraft-flask/issues/20